versioned/storage/common/src/test/java/org/projectnessie/versioned/storage/common/logic/TestMergeBase.java

```diff
@@ -0,0 +1,438 @@
+                .respectMergeParents(false)
+                .targetCommitId(e.id())
+                .fromCommitId(f.id())
+                .build()
+                .identifyMergeBase())
+        .isEqualTo(a);
+  }
+
+  /**
+   * Merge-again case. <code><pre>
+   *              ----E----G---------I
+   *            /       /           /
+   *       ----B----D----------H----
+   *      /               /
+   * ----A------C--------F
+   * </pre></code>
+   *
+   * <p>Best merge-base of {@code I} onto {@code F} is {@code F}.
+   */
+  @Test
+  void doubleMerge3() {
+    CommitObj a = repo.add(initialCommit());
+    CommitObj b = repo.add(buildCommit("b").addTail(a.id()));
+    CommitObj c = repo.add(buildCommit("c").addTail(a.id()));
+    CommitObj d = repo.add(buildCommit("d").addTail(b.id()));
+    CommitObj e = repo.add(buildCommit("e").addTail(b.id()));
+    CommitObj f = repo.add(buildCommit("f").addTail(c.id()));
+    CommitObj g = repo.add(buildCommit("g").addTail(e.id()).addSecondaryParents(d.id()));
+    CommitObj h = repo.add(buildCommit("h").addTail(d.id()).addSecondaryParents(f.id()));
+    CommitObj i = repo.add(buildCommit("i").addTail(g.id()).addSecondaryParents(h.id()));
+
+    soft.assertThat(
+            MergeBase.builder()
+                .loadCommit(repo::loadCommit)
+                .targetCommitId(f.id())
+                .fromCommitId(i.id())
+                .build()
+                .identifyMergeBase())
+        .isEqualTo(f);
+    soft.assertThat(
+            MergeBase.builder()
+                .loadCommit(repo::loadCommit)
+                .respectMergeParents(false)
+                .targetCommitId(f.id())
+                .fromCommitId(i.id())
+                .build()
+                .identifyMergeBase())
+        .isEqualTo(a);
+  }
+
+  /**
+   * Cross-merge case. <code><pre>
+   *        ----B--------D----F
+   *       /       \ /
+   *      /        / \
+   * ----A------C--------E----G
+   * </pre></code>
+   *
+   * <p>Merge-base outcome for {@code F} onto {@code G} is {@code B}. In theory, it could also be
+   * {@code C}, but the logic/implementation ensures that it's always {@code B}.
+   */
+  @Test
+  void afterCrossMerge() {
+    CommitObj a = repo.add(initialCommit());
+    CommitObj b = repo.add(buildCommit("b").addTail(a.id()));
+    CommitObj c = repo.add(buildCommit("c").addTail(a.id()));
+    CommitObj d = repo.add(buildCommit("d").addTail(b.id()).addSecondaryParents(c.id()));
+    CommitObj e = repo.add(buildCommit("e").addTail(c.id()).addSecondaryParents(b.id()));
+    CommitObj f = repo.add(buildCommit("f").addTail(d.id()));
+    CommitObj g = repo.add(buildCommit("g").addTail(e.id()));
+
+    soft.assertThat(
+            MergeBase.builder()
+                .loadCommit(repo::loadCommit)
+                .targetCommitId(g.id())
+                .fromCommitId(f.id())
+                .build()
+                .identifyMergeBase())
+        .isEqualTo(b);
```

Comment:
Only `B` will be selected here, because of the `.get(0)` in the last line of `MergeBase.identifyMergeBase()`, which chooses the "best" merge base.

We should however add an ability to let a user chose the "right/best" merge-base - as a follow-up though. Not sure whether that particular test case is a legit prod use case anyway. Manually choosing the "right/best" merge-base is difficult for humans in complex graphs anyway. I cannot remember a case when I had to force Git to use a specific merge-base.

Reply:
See added VS-test cases